osu.Game/Overlays/NewChat/ChannelListingItem.cs

```diff
@@ -0,0 +1,175 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Online.Chat;
+using osuTK;
+
+namespace osu.Game.Overlays.NewChat
+{
+    public class ChannelListingItem : OsuClickableContainer, IFilterable
+    {
+        public event Action<Channel>? OnRequestJoin;
+        public event Action<Channel>? OnRequestLeave;
+
+        public bool FilteringActive { get; set; }
+        public IEnumerable<string> FilterTerms => new[] { channel.Name, channel.Topic ?? string.Empty };
+        public bool MatchingFilter
+        {
+            set => this.FadeTo(value ? 1f : 0f, 100);
+        }
+
+        private readonly float TEXT_SIZE = 18;
+        private readonly float ICON_SIZE = 14;
+        private readonly Channel channel;
+
+        private Colour4 selectedColour;
+        private Colour4 normalColour;
+
+        private Box hoverBox = null!;
+        private SpriteIcon checkbox = null!;
+        private OsuSpriteText channelText = null!;
+        private IBindable<bool> channelJoined = null!;
+
+        [Resolved]
+        private OverlayColourProvider overlayColours { get; set; } = null!;
+
+        public ChannelListingItem(Channel channel)
+        {
+            this.channel = channel;
+
+            Masking = true;
+            CornerRadius = 5;
+            RelativeSizeAxes = Axes.X;
+            Height = 20;
+        }
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            hoverBox.Show();
+            return base.OnHover(e);
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            hoverBox.Hide();
+            base.OnHoverLost(e);
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            // Set colours
+            normalColour = overlayColours.Light3;
+            selectedColour = Colour4.White;
+
+            // Set handlers for state display
+            channelJoined = channel.Joined.GetBoundCopy();
+            channelJoined.BindValueChanged(change =>
+            {
+                if (change.NewValue)
+                {
+                    checkbox.Show();
+                    channelText.Colour = selectedColour;
+                }
+                else
+                {
+                    checkbox.Hide();
+                    channelText.Colour = normalColour;
+                }
+            }, true);
+
+            // Set action on click
+            Action = () => (channelJoined.Value ? OnRequestLeave : OnRequestJoin)?.Invoke(channel);
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```

Comment:
order all over the place here too - preferred ordering is:

* ctor
* BDL
* `LoadComplete()`
* event handlers/protected methods
* private methods

Also see: https://github.com/ppy/osu-framework/issues/2278

Reply:
I didn't know about this ordering guide. I will make sure to use it.